### ISSUE TITLE
Valid permissions object path error when forward slashes in schema names

### ISSUE THREAD

**Issue by Alxander64** [metabase/metabase#12450]
**Describe the bug**
Hi, I think I've run into an issue similar to #8693, except with forward slashes. I'm using Metabase with Presto, and the catalog has schemas which include forward slashes in their names.

Unfortunately I can't change this. For context, I'm using Presto with an Apache Pulsar connector plugin, and Pulsar itself names topics as `tenant/namespace/topic`, where `tenant/namespace` appear to be the schemas in Presto. Metabase works with Presto if I do native queries of the form

```
select * from pulsar."public/default".topic_name
```

but this issue is preventing me from discovering data as one normally would in Metabase, since it's having trouble finding the tables on its own.

**Logs**
Here are the logs from the Docker container.

```
05-03 20:27:48 ERROR middleware.log :: GET /api/database/2/schemas 500 3.1 ms (4 DB calls) 
{:value "/db/2/schema/public/functions/",
 :error (not ("Valid permissions object path." a-java.lang.String)),
 :message "Output of object-path does not match schema: \n\n\t   (not (\"Valid permissions object path.\" a-java.lang.String))  \n\n",
 :type clojure.lang.ExceptionInfo,
 :stacktrace
 ["--> models.permissions$fn__30618$object_path__30631.invoke(permissions.clj:124)"
  "api.database$can_read_schema_QMARK_.invokeStatic(database.clj:637)"
  "api.database$can_read_schema_QMARK_.invoke(database.clj:632)"
  "api.database$fn__54640.invokeStatic(database.clj:643)"
  "api.database$fn__54640.invoke(database.clj:639)"
  "middleware.auth$enforce_authentication$fn__66808.invoke(auth.clj:14)"
  "routes$fn__68332$fn__68333.doInvoke(routes.clj:56)"
  "middleware.exceptions$catch_uncaught_exceptions$fn__66907.invoke(exceptions.clj:97)"
  "middleware.exceptions$catch_api_exceptions$fn__66904.invoke(exceptions.clj:85)"
  "middleware.log$log_api_call$fn__68710$fn__68711.invoke(log.clj:197)"
  "middleware.log$log_api_call$fn__68710.invoke(log.clj:191)"
  "middleware.security$add_security_headers$fn__66870.invoke(security.clj:121)"
  "middleware.json$wrap_json_body$fn__68413.invoke(json.clj:64)"
  "middleware.json$wrap_streamed_json_response$fn__68431.invoke(json.clj:100)"
  "middleware.misc$maybe_set_site_url$fn__68740.invoke(misc.clj:59)"
  "middleware.session$bind_current_user$fn__63397$fn__63398.invoke(session.clj:209)"
  "middleware.session$do_with_current_user.invokeStatic(session.clj:184)"
  "middleware.session$do_with_current_user.invoke(session.clj:177)"
  "middleware.session$bind_current_user$fn__63397.invoke(session.clj:208)"
  "middleware.session$wrap_current_user_id$fn__63382.invoke(session.clj:161)"
  "middleware.session$wrap_session_id$fn__63326.invoke(session.clj:125)"
  "middleware.auth$wrap_api_key$fn__66816.invoke(auth.clj:27)"
  "middleware.misc$bind_user_locale$fn__68745.invoke(misc.clj:78)"
  "middleware.misc$add_content_type$fn__68726.invoke(misc.clj:29)"
  "middleware.misc$disable_streaming_buffering$fn__68754.invoke(misc.clj:94)"
  "server$async_proxy_handler$fn__68488.invoke(server.clj:70)"
  "server.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)"]}
```

I also get this message on screen when browsing the database:

```
Output of object-path does not match schema: [0;33m (not ("Valid permissions object path." a-java.lang.String)) [0m
```

**To Reproduce**
1. Connect a Presto database to Metabase, where schemas include forward slashes in their names
2. Browse the database on Metabase
3. See logs from Metabase

**Expected behavior**
Would like to be able to explore data, as one normally would with the Sample Dataset for instance.

**Information about your Metabase Installation:**

From: Admin -> Troubleshooting.
I am also running Metabase in Docker.

```
{
  "browser-info": {
    "language": "en-GB",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.129 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.5+10",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.5",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.5+10",
    "os.name": "Linux",
    "os.version": "5.3.0-51-generic",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "presto"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "application-database-details": {
      "database": {
        "name": "H2",
        "version": "1.4.197 (2018-03-18)"
      },
      "jdbc-driver": {
        "name": "H2 JDBC Driver",
        "version": "1.4.197 (2018-03-18)"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-04-21",
      "tag": "v0.35.3",
      "branch": "release-0.35.x",
      "hash": "1d424cb"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**
Blocking usage of data discovery features in Metabase for the affected database.

**Comment by Alxander64** [comment]
I noticed in the commit of #8889, this was commented as a TODO.

For the second capture group used in the regex's that were initially edited, would using `([a-zA-Z0-9/]*)` work in this case?

For example, instead of:

```
#"^/db/(\d+)/schema/([^/]*)/$"
#"^/db/(\d+)/schema/([^/]*)/table/(\d+)/$"
#"^/db/(\d+)/schema/([^/]*)/table/(\d+)/read/$"
#"^/db/(\d+)/schema/([^/]*)/table/(\d+)/query/$"
```

We have:

```
#"^/db/(\d+)/schema/([a-zA-Z0-9/]*)/$"
#"^/db/(\d+)/schema/([a-zA-Z0-9/]*)/table/(\d+)/$"
#"^/db/(\d+)/schema/([a-zA-Z0-9/]*)/table/(\d+)/read/$"
#"^/db/(\d+)/schema/([a-zA-Z0-9/]*)/table/(\d+)/query/$"
```

I was just playing around in regexr and found that this second capture group correctly captures `public/default` as the schema, so it seems to work at least in my case.

I can try to run tests with this and create a pull request, just wondering if there are any objections to the idea that I'm not seeing, I've only just recently started using Metabase and don't know all too much about it.

**Comment by Alxander64** [comment]
Actually, I don't see why the second capture group can't just be `(.*)`, simply just removing the negated set `[^/]` and accepting any characters. I'm not sure if this will affect any other part of the application indirectly, but I will go ahead and try testing this myself.

**Comment by Alxander64** [comment]
Tested the changes I stated above, `lein test` ended with the following

```
Ran 3325 tests containing 6974 assertions.
17 failures, 0 errors.
Tests failed.
Error encountered performing task 'test' with profile(s): 'base,system,user,provided,dev,test'
Tests failed.
```

I'm not all too familiar with testing Clojure code, so I'm not sure where to look from here.

**Comment by camsaul** [comment]
To fix this we just need to update the regexes in `metabase.models.permissions`. We can remove the capture groups since we don't use them. 

I actually got 90% of the way thru fixing this at one point and IIRC I used some lookbehind/lookahead magic to make sure it worked as expected. Trying to figure out where that code went

**Comment by camsaul** [comment]
Update: I think I actually fixed this in https://github.com/metabase/metabase-enterprise/pull/502/commits/07f6649849270458f7a683ccab2da438c961e353 but haven't backported the fix to OSS yet. I ended up just escaping forward slashes in the permissions paths to fix stuff. 

For this fix we can just cherry-pick the changes to `metabase.models.permissions` and `metabase.models.permissions-test` and add some extra end-to-end tests that sync a database with a slash in a table/field name and make sure we can actually query it

**Comment by camsaul** [comment]
Fixed by  #13476
